apis/compute/v1alpha1/computeregiontargettcpproxy_types.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/apis/k8s/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var ComputeRegionTargetTCPProxyGVK = GroupVersion.WithKind("ComputeRegionTargetTCPProxy")
+
+// ComputeRegionTargetTCPProxySpec defines the desired state of ComputeRegionTargetTCPProxy
+// +kcc:proto=google.cloud.compute.v1.TargetTcpProxy
+type ComputeRegionTargetTCPProxySpec struct {
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="BackendServiceRef is immutable"
+	// Immutable. A reference to the ComputeBackendService resource.
+	// +required
+	BackendServiceRef *refs.ComputeBackendServiceRef `json:"backendServiceRef"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Description is immutable"
+	// Immutable. An optional description of this resource.
+	Description *string `json:"description,omitempty"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Region is immutable"
+	// Immutable. The geographical location of the ComputeTargetTCPProxy.
+	// Reference: GCP definition of regions/zones (https://cloud.google.com/compute/docs/regions-zones/)
+	Region *string `json:"region"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="ProxyBind is immutable"
+	// Immutable. This field only applies when the forwarding rule that references
+	// this target proxy has a loadBalancingScheme set to INTERNAL_SELF_MANAGED.
+	ProxyBind *bool `json:"proxyBind,omitempty"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="ProxyHeader is immutable"
+	// Immutable. Specifies the type of proxy header to append before sending data to
+	// the backend. Default value: "NONE" Possible values: ["NONE", "PROXY_V1"].
+	ProxyHeader *string `json:"proxyHeader,omitempty"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="ResourceID field is immutable"
+	// Immutable.
```

Comment:
suggest not checking the immutability. User can specify `spec.resourceID` if it refers to the same obj. But I realized it is too trivial to capture all these conditions, and it further confuses the users.

Reply:
ResourceID is an immutable field in the existing types: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/6e9579347aadd08c07cfb0f1bd1747c4c9f4b197/pkg/clients/generated/apis/compute/v1beta1/computetargettcpproxy_types.go#L56. During direct controller migration I prefer to not change this behavior.